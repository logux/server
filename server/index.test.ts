import spawn from 'cross-spawn'
import type { ChildProcess, SpawnOptions } from 'node:child_process'
import { join } from 'node:path'
import { afterEach, expect, it } from 'vitest'

import { Server } from '../index.js'

const ROOT = join(import.meta.dirname, '..')
const DATE = /\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d/g

let started: ChildProcess | undefined

function start(name: string, args?: string[]): Promise<void> {
  return new Promise<void>(resolve => {
    started = spawn(join(ROOT, 'test/servers/', name), args)
    let running = false
    function callback(): void {
      if (!running) {
        running = true
        resolve()
      }
    }
    started.stdout?.on('data', callback)
    started.stderr?.on('data', callback)
  })
}

function check(
  name: string,
  args?: string[],
  opts?: SpawnOptions,
  kill = false
): Promise<[string, number]> {
  return new Promise<[string, number]>(resolve => {
    let out = ''
    let server = spawn(join(ROOT, 'test/servers/', name), args, opts)
    server.stdout?.on('data', chank => {
      out += chank
    })
    server.stderr?.on('data', chank => {
      out += chank
    })
    server.on('close', exit => {
      let fixed = out
        .replace(/[^\n]+DeprecationWarning[^\n]+\n/gm, '')
        .replace(DATE, '1970-01-01 00:00:00')
        .replace(/"time":"[^"]+"/g, '"time":"1970-01-01T00:00:00.000Z"')
        .replace(/PID:(\s+)\d+/, 'PID:$121384')
        .replace(/"pid":\d+,/g, '"pid":21384,')
        .replace(/Logux server:( +)\d+.\d+.\d+/g, 'Logux server:$10.0.0')
        .replace(/"loguxServer":"\d+.\d+.\d+"/g, '"loguxServer":"0.0.0"')
        .replace(/"hostname":"[^"]+"/g, '"hostname":"localhost"')
      fixed = fixed.replace(/\r\v/g, '\n')
      resolve([fixed, exit || 0])
    })

    function waitOut(): void {
      if (out.length > 0) {
        server.kill('SIGINT')
      } else {
        setTimeout(waitOut, 500)
      }
    }
    if (kill) setTimeout(waitOut, 500)
  })
}

function fakeProcess(argv: string[] = [], env: object = {}): any {
  return { argv, env }
}

async function checkOut(
  name: string,
  args?: string[],
  opts?: SpawnOptions
): Promise<void> {
  let result = await check(name, args, opts, true)
  let out = result[0]
  let exit = result[1]
  expect(out).toMatchSnapshot()
  if (exit !== 0) {
    throw new Error(`Fall with:\n${out}`)
  }
}

async function checkError(
  name: string,
  args?: string[],
  opts?: SpawnOptions
): Promise<void> {
  let result = await check(name, args, opts)
  let out = result[0]
  let exit = result[1]
  expect(out).toMatchSnapshot()
  expect(exit).toEqual(1)
}

afterEach(() => {
  if (started) {
    started.kill('SIGINT')
    started = undefined
  }
})

it('uses CLI args for options', () => {
  let options = Server.loadOptions(
    fakeProcess([
      '',
      '--port',
      '1337',
      '--host',
      '192.168.1.1',
      '--logger',
      'json',
      '--redis',
      '//localhost'
    ]),
    {
      minSubprotocol: 1,
      subprotocol: 1
    }
  )

  expect(options.host).toEqual('192.168.1.1')
  expect(options.port).toEqual(1337)
  expect(options.logger).toEqual('json')
  expect(options.cert).toBeUndefined()
  expect(options.key).toBeUndefined()
  expect(options.redis).toEqual('//localhost')
})

it('uses env for options', () => {
  let options = Server.loadOptions(
    fakeProcess([], {
      LOGUX_HOST: '127.0.1.1',
      LOGUX_LOGGER: 'json',
      LOGUX_PORT: '31337',
      LOGUX_REDIS: '//localhost'
    }),
    {
      minSubprotocol: 1,
      subprotocol: 1
    }
  )

  expect(options.host).toEqual('127.0.1.1')
  expect(options.port).toEqual(31337)
  expect(options.logger).toEqual('json')
  expect(options.redis).toEqual('//localhost')
})

it('uses combined options', () => {
  let options = Server.loadOptions(
    fakeProcess(['', '--key', './key.pem'], { LOGUX_CERT: './cert.pem' }),
    { minSubprotocol: 1, port: 31337, subprotocol: 1 }
  )

  expect(options.port).toEqual(31337)
  expect(options.cert).toEqual('./cert.pem')
  expect(options.key).toEqual('./key.pem')
})

it('uses arg, env, options in given priority', () => {
  let options1 = Server.loadOptions(
    fakeProcess(['', '--port', '31337'], { LOGUX_PORT: 21337 }),
    {
      minSubprotocol: 1,
      port: 11337,
      subprotocol: 1
    }
  )
  let options2 = Server.loadOptions(fakeProcess([], { LOGUX_PORT: 21337 }), {
    minSubprotocol: 1,
    port: 11337,
    subprotocol: 1
  })
  let options3 = Server.loadOptions(fakeProcess(), {
    minSubprotocol: 1,
    port: 11337,
    subprotocol: 1
  })

  expect(options1.port).toEqual(31337)
  expect(options2.port).toEqual(21337)
  expect(options3.port).toEqual(11337)
})

it('destroys everything on exit', () => checkOut('destroy.js'))

it('writes about unbind', async () => {
  let result = await check('unbind.js', [], {}, true)
  expect(result[0]).toMatchSnapshot()
})

it('shows uncatch errors', () => checkError('throw.js'))

it('shows uncatch rejects', () => checkError('uncatch.js'))

it('uses environment variables for config', () => {
  return checkOut('options.js', [], {
    env: {
      ...process.env,
      LOGUX_LOGGER: 'json',
      LOGUX_PORT: '31337',
      NODE_ENV: 'test'
    }
  })
})

it('uses logger param', () => checkOut('options.js', ['', '-l', 'json']))

it('uses autoload modules', () => checkOut('autoload-modules.js'))

it('uses autoload wrong export', () => checkError('autoload-error-modules.js'))

it('uses .env cwd', async () => {
  let result = await check(
    'options.js',
    [],
    { cwd: join(ROOT, 'test/fixtures') },
    true
  )
  expect(result[0]).toMatchSnapshot()
})

it('uses .env from root', () => checkOut('root.js'))

it('shows help', async () => {
  await checkOut('options.js', ['', '--help'], {
    env: {
      ...process.env,
      NO_COLOR: '1'
    }
  })
})

it('shows help about port in use', async () => {
  await start('eaddrinuse.js')
  let result = await check('eaddrinuse.js')

  expect(result[0]).toMatchSnapshot()
})

it('shows help about privileged port', () => checkError('eacces.js'))

it('shows help about unknown option', () => checkError('unknown.js'))

it('shows help about missed option', () => checkError('missed.js'))

it('disables colors for constructor errors', () => {
  return checkError('missed.js', [], {
    env: {
      ...process.env,
      NODE_ENV: 'production'
    }
  })
})

it('uses logger param for constructor errors', () => {
  return checkError('missed.js', ['', '-l', 'json'])
})

it('writes JSON log', () => checkOut('json.js'))

it('has custom logger', () => checkOut('logger.js'))
