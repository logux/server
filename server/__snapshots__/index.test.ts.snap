// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`destroys everything on exit 1`] = `
" INFO   Logux server is listening at 1970-01-01 00:00:00
        PID:          21384
        Logux server: 0.0.0
        Environment:  test
        Node ID:      server:FnXaqDxY
        Subprotocol:  1.0.0
        Supports:     1.x
        Health check: http://127.0.0.1:2000/health
        Listen:       ws://127.0.0.1:2000/

 INFO   Shutting down Logux server at 1970-01-01 00:00:00

 INFO   Custom destroy task finished at 1970-01-01 00:00:00

"
`;

exports[`disables colors for constructor errors 1`] = `
" FATAL  Missed supports option in server constructor at 1970-01-01 00:00:00
        Check server constructor and Logux Server documentation

"
`;

exports[`has custom logger 1`] = `
" INFO   Hi from custom logger at 1970-01-01 00:00:00
        Field: 1

 INFO   Logux server is listening at 1970-01-01 00:00:00
        PID:          21384
        Logux server: 0.0.0
        Environment:  test
        Node ID:      server:FnXaqDxY
        Subprotocol:  1.0.0
        Supports:     1.x
        Health check: http://127.0.0.1:31337/health
        Listen:       ws://127.0.0.1:31337/

 INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`shows help 1`] = `
"Start Logux Server

Options:
--host            LOGUX_HOST            Host to bind server
--port            LOGUX_PORT            Port to bind server
--key             LOGUX_KEY             Path to SSL key
--cert            LOGUX_CERT            Path to SSL certificate
--supports        LOGUX_SUPPORTS        Range of supported client subprotocols
--subprotocol     LOGUX_SUBPROTOCOL     Server subprotocol
--logger          LOGUX_LOGGER          Logger type
--backend         LOGUX_BACKEND         Backend to process actions
--control-secret  LOGUX_CONTROL_SECRET  Secret to control Logux server
--control-mask    LOGUX_CONTROL_MASK    CIDR masks of control servers
--redis           LOGUX_REDIS           Redis URL for Logux Server Pro scaling

Examples:
options.js --port 31337 --host 127.0.0.1
LOGUX_PORT=1337 LOGUX_HOST=127.0.0.1 options.js
"
`;

exports[`shows help about missed option 1`] = `
" FATAL  Missed supports option in server constructor at 1970-01-01 00:00:00
        Check server constructor and Logux Server documentation

"
`;

exports[`shows help about missed secret 1`] = `
" FATAL  backend requires controlSecret option at 1970-01-01 00:00:00
        Call npx nanoid and set result as controlSecret or LOGUX_CONTROL_SECRET
        environment variable

"
`;

exports[`shows help about port in use 1`] = `
" FATAL  Port 2001 already in use at 1970-01-01 00:00:00
        Another Logux server or other app already running on this port.
        Probably you haven’t stopped server from other project or previous
        version of this server was not killed.
        
        $ su - root
        # netstat -nlp | grep 2001
        Proto   Local Address   State    PID/Program name
        tcp     0.0.0.0:2001    LISTEN   777/node
        # sudo kill -9 777

 INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`shows help about privileged port 1`] = `
" FATAL  You are not allowed to run server on port 1000 at 1970-01-01 00:00:00
        Non-privileged users can't start a listening socket on ports below 1024.
        Try to change user or take another port.
        
        $ su - <username>
        $ npm start -p 1000

 INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`shows help about unknown option 1`] = `
" FATAL  Missed supports option in server constructor at 1970-01-01 00:00:00
        Check server constructor and Logux Server documentation

"
`;

exports[`shows uncatch errors 1`] = `
" FATAL  Test Error at 1970-01-01 00:00:00
        fake stacktrace

 INFO   Shutting down Logux server at 1970-01-01 00:00:00

 INFO   Fatal event: Test Error at 1970-01-01 00:00:00

"
`;

exports[`shows uncatch rejects 1`] = `
" FATAL  Test Error at 1970-01-01 00:00:00
        fake stacktrace

 INFO   Shutting down Logux server at 1970-01-01 00:00:00

 INFO   Fatal event: Test Error at 1970-01-01 00:00:00

"
`;

exports[`uses .env cwd 1`] = `
" INFO   Logux server is listening at 1970-01-01 00:00:00
        PID:          21384
        Logux server: 0.0.0
        Environment:  test
        Node ID:      server:FnXaqDxY
        Subprotocol:  1.0.0
        Supports:     1.x
        Health check: http://127.0.0.1:3334/health
        Listen:       ws://127.0.0.1:3334/

 INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`uses .env from root 1`] = `
" INFO   Logux server is listening at 1970-01-01 00:00:00
        PID:          21384
        Logux server: 0.0.0
        Environment:  test
        Node ID:      server:FnXaqDxY
        Subprotocol:  1.0.0
        Supports:     1.x
        Health check: http://127.0.0.1:3334/health
        Listen:       ws://127.0.0.1:3334/

 INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`uses autoload modules 1`] = `
"Root path module: 1.0.0
Child path module: 1.0.0
 INFO   Logux server is listening at 1970-01-01 00:00:00
        PID:          21384
        Logux server: 0.0.0
        Environment:  test
        Node ID:      server:FnXaqDxY
        Subprotocol:  1.0.0
        Supports:     1.x
        Health check: http://127.0.0.1:31337/health
        Listen:       ws://127.0.0.1:31337/

 INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`uses autoload wrong export 1`] = `
" INFO   Logux server is listening at 1970-01-01 00:00:00
        PID:          21384
        Logux server: 0.0.0
        Environment:  test
        Node ID:      server:FnXaqDxY
        Subprotocol:  1.0.0
        Supports:     1.x
        Health check: http://127.0.0.1:31337/health
        Listen:       ws://127.0.0.1:31337/

 FATAL  Server module should has default export with function that accepts a server at 1970-01-01 00:00:00
        error-modules/wrond-export/index.js default export is string

 INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`uses environment variables for config 1`] = `
"{\\"level\\":30,\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"pid\\":21384,\\"hostname\\":\\"localhost\\",\\"name\\":\\"logux-server\\",\\"loguxServer\\":\\"0.0.0\\",\\"environment\\":\\"test\\",\\"nodeId\\":\\"server:FnXaqDxY\\",\\"subprotocol\\":\\"1.0.0\\",\\"supports\\":\\"1.x\\",\\"listen\\":\\"ws://127.0.0.1:31337/\\",\\"healthCheck\\":\\"http://127.0.0.1:31337/health\\",\\"msg\\":\\"Logux server is listening\\"}
{\\"level\\":30,\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"pid\\":21384,\\"hostname\\":\\"localhost\\",\\"name\\":\\"logux-server\\",\\"msg\\":\\"Shutting down Logux server\\"}
"
`;

exports[`uses logger param 1`] = `
"{\\"level\\":30,\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"pid\\":21384,\\"hostname\\":\\"localhost\\",\\"name\\":\\"logux-server\\",\\"loguxServer\\":\\"0.0.0\\",\\"environment\\":\\"test\\",\\"nodeId\\":\\"server:FnXaqDxY\\",\\"subprotocol\\":\\"1.0.0\\",\\"supports\\":\\"1.x\\",\\"listen\\":\\"ws://127.0.0.1:31337/\\",\\"healthCheck\\":\\"http://127.0.0.1:31337/health\\",\\"msg\\":\\"Logux server is listening\\"}
{\\"level\\":30,\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"pid\\":21384,\\"hostname\\":\\"localhost\\",\\"name\\":\\"logux-server\\",\\"msg\\":\\"Shutting down Logux server\\"}
"
`;

exports[`uses logger param for constructor errors 1`] = `
"{\\"level\\":60,\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"pid\\":21384,\\"hostname\\":\\"localhost\\",\\"name\\":\\"logux-server\\",\\"note\\":\\"Check server constructor and Logux Server documentation\\",\\"msg\\":\\"Missed \`supports\` option in server constructor\\"}
"
`;

exports[`writes about unbind 1`] = `
" INFO   Shutting down Logux server at 1970-01-01 00:00:00

"
`;

exports[`writes to custom pino log 1`] = `
"{\\"level\\":30,\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"pid\\":21384,\\"hostname\\":\\"localhost\\",\\"name\\":\\"logux-server-custom\\",\\"customProp\\":\\"42\\",\\"loguxServer\\":\\"0.0.0\\",\\"environment\\":\\"test\\",\\"nodeId\\":\\"server:FnXaqDxY\\",\\"subprotocol\\":\\"1.0.0\\",\\"supports\\":\\"1.x\\",\\"listen\\":\\"ws://127.0.0.1:2000/\\",\\"healthCheck\\":\\"http://127.0.0.1:2000/health\\",\\"msg\\":\\"Logux server is listening\\"}
{\\"level\\":30,\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"pid\\":21384,\\"hostname\\":\\"localhost\\",\\"name\\":\\"logux-server-custom\\",\\"customProp\\":\\"42\\",\\"msg\\":\\"Shutting down Logux server\\"}
"
`;

exports[`writes to pino log 1`] = `
"{\\"level\\":30,\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"pid\\":21384,\\"hostname\\":\\"localhost\\",\\"name\\":\\"logux-server\\",\\"loguxServer\\":\\"0.0.0\\",\\"environment\\":\\"test\\",\\"nodeId\\":\\"server:FnXaqDxY\\",\\"subprotocol\\":\\"1.0.0\\",\\"supports\\":\\"1.x\\",\\"listen\\":\\"ws://127.0.0.1:2000/\\",\\"healthCheck\\":\\"http://127.0.0.1:2000/health\\",\\"msg\\":\\"Logux server is listening\\"}
{\\"level\\":30,\\"time\\":\\"1970-01-01T00:00:00.000Z\\",\\"pid\\":21384,\\"hostname\\":\\"localhost\\",\\"name\\":\\"logux-server\\",\\"msg\\":\\"Shutting down Logux server\\"}
"
`;
